src/context/blocked.rs

```diff
@@ -1,18 +1,177 @@
-use std::ptr;
+use crate::redismodule::{AUTH_HANDLED, AUTH_NOT_HANDLED};
+use crate::{raw, Context, ValkeyError, ValkeyString};
+use std::os::raw::{c_int, c_void};
 
-use crate::raw;
-use crate::Context;
+// Callback types for handling blocked client operations
+#[derive(Debug)]
+pub enum ReplyCallback<T> {
+    Auth(fn(&Context, ValkeyString, ValkeyString, T) -> Result<c_int, ValkeyError>),
+}
+
+// Callback type for freeing private data associated with a blocked client
+type FreePrivDataCallback<T> = fn(&Context, T);
 
-pub struct BlockedClient {
+pub struct BlockedClient<T = ()> {
     pub(crate) inner: *mut raw::RedisModuleBlockedClient,
+    reply_callback: Option<ReplyCallback<T>>,
+    free_priv_callback: Option<FreePrivDataCallback<T>>,
+    private_data: Option<Box<T>>,
+}
+
+#[allow(dead_code)]
+unsafe extern "C" fn auth_reply_wrapper<T: 'static + Clone>(
+    ctx: *mut raw::RedisModuleCtx,
+    username: *mut raw::RedisModuleString,
+    password: *mut raw::RedisModuleString,
+    err: *mut *mut raw::RedisModuleString,
+) -> c_int {
+    let context = Context::new(ctx);
+    let ctx_ptr = std::ptr::NonNull::new_unchecked(ctx);
+    let username = ValkeyString::new(Some(ctx_ptr), username);
+    let password = ValkeyString::new(Some(ctx_ptr), password);
+
+    // Get private data from context
+    let priv_data = context.get_blocked_client_privdata();
+    if priv_data.is_null() {
+        context.log_warning("[auth_reply_wrapper] Private data is null; this should not happen!");
+        return AUTH_NOT_HANDLED;
+    }
+
+    let callbacks =
+        &*(priv_data as *const (ReplyCallback<T>, Option<FreePrivDataCallback<T>>, Box<T>));
+
+    match &callbacks.0 {
+        ReplyCallback::Auth(cb) => {
+            match cb(&context, username, password, callbacks.2.as_ref().clone()) {
+                Ok(status) => status,
+                Err(error) => {
+                    let error_msg = ValkeyString::create_and_retain(&error.to_string());
+                    *err = error_msg.inner;
+                    AUTH_HANDLED
+                }
+            }
+        }
+    }
+}
+
+#[allow(dead_code)]
+unsafe extern "C" fn free_callback_wrapper<T: 'static>(
+    ctx: *mut raw::RedisModuleCtx,
+    privdata: *mut c_void,
+) {
+    let context = Context::new(ctx);
+
+    if privdata.is_null() {
+        context
+            .log_warning("[free_callback_wrapper] Private data is null; this should not happen!");
+        return;
+    }
+
+    let (_, free_cb, data) = *Box::from_raw(
```

Comment:
While the tuple containing 3 elements works, IMO it is slightly more error prone than using a dedicated structure that defines the elements that priv data should contain.

What do you think?